libs/core/type_support/include/hpx/type_support/uninitialized_relocate.hpp

```diff
@@ -0,0 +1,162 @@
+                    can_move_construct_nothrow ? 1 : // for loop nothrow
+                                                 2>; // for loop try-catch
+
+            constexpr static int value = type::value;
+        };
+
+        template <typename InIter, typename FwdIter,
+            std::enable_if_t<choose_uninitialized_relocate_helper<InIter,
+                                 FwdIter>::value == 0,
+                int> = 0>
+        FwdIter uninitialized_relocate_helper(
+            InIter first, InIter last, FwdIter dst) noexcept
+        {
+            auto n_objects = std::distance(first, last);
+
+            if (n_objects != 0)
+            {
+                std::byte* first_byte =
+                    reinterpret_cast<std::byte*>(std::addressof(*first));
+                std::byte* last_byte =
+                    reinterpret_cast<std::byte*>(std::addressof(*last));
+
+                auto n_bytes = std::distance(first_byte, last_byte);
+
+                void* dst_void = static_cast<void*>(std::addressof(*dst));
+                void* first_void = static_cast<void*>(std::addressof(*first));
+
+                std::memmove(dst_void, first_void, n_bytes);
+
+                dst += n_objects;
+            }
+
+            return dst;
+        }
+
+        template <typename InIter, typename FwdIter,
+            std::enable_if_t<choose_uninitialized_relocate_helper<InIter,
+                                 FwdIter>::value == 1,
+                int> = 0>
+        FwdIter uninitialized_relocate_helper(
+            InIter first, InIter last, FwdIter dst) noexcept
+        {
+            for (; first != last; ++first, ++dst)
+            {
+                // the move + destroy version will be used
+                hpx::relocate_at(std::addressof(*first), std::addressof(*dst));
+            }
+
+            return dst;
+        }
+
+        template <typename InIter, typename FwdIter,
+            std::enable_if_t<choose_uninitialized_relocate_helper<InIter,
+                                 FwdIter>::value == 2,
+                int> = 0>
+        FwdIter uninitialized_relocate_helper(
+            InIter first, InIter last, FwdIter dst)
+        {
+            FwdIter original_dst = dst;
+
+            for (; first != last; ++first, ++dst)
+            {
+                try
+                {
+                    // the move + destroy version will be used
+                    hpx::relocate_at(
+                        std::addressof(*first), std::addressof(*dst));
+                }
+                catch (...)
+                {
+                    // destroy all objects other that the one
+                    // that caused the exception
+
+                    // destroy all objects constructed so far
+                    std::destroy(original_dst, dst);
+                    // destroy all the objects not relocated yet
+                    std::destroy(++first, last);
+
+                    throw;
+                }
+            }
```

Comment:
We do have `loop_with_cleanup` which could be used here (see: https://github.com/STEllAR-GROUP/hpx/blob/dd17010adf02cdab681d96f10f96626d476459c0/libs/core/algorithms/include/hpx/parallel/util/loop.hpp#L625-L631)